File: javascript/src/verification.js
/**
 * Credential verification for AgentPin.
 * Implements the 12-step verification flow from the spec.
 */

import { decodeJwtUnverified, verifyJwt } from './jwt.js';
import { jwkToPem } from './jwk.js';
import { validateCredentialAgainstDiscovery } from './credential.js';
import { validateDiscoveryDocument, findKeyByKid, findAgentById, fetchDiscoveryDocument } from './discovery.js';
import { checkRevocation, fetchRevocationDocument } from './revocation.js';
import { checkPinning, PinningResult } from './pinning.js';
import { constraintsSubsetOf } from './constraint.js';
import { ErrorCode, AgentPinError } from './types.js';

/**
 * Default verifier configuration.
 * @returns {{ clockSkewSecs: number, maxTtlSecs: number }}
 */
export function defaultVerifierConfig() {
    return {
        clockSkewSecs: 60,
        maxTtlSecs: 86400,
    };
}

/**
 * Create a successful verification result.
 */
function successResult(agentId, issuer, capabilities, constraints, pinStatus) {
    return {
        valid: true,
        agent_id: agentId,
        issuer,
        capabilities,
        constraints: constraints || null,
        delegation_verified: null,
        delegation_chain: null,
        key_pinning: pinStatus,
        error_code: null,
        error_message: null,
        warnings: [],
    };
}

/**
 * Create a failed verification result.
 */
function failureResult(code, message) {
    return {
        valid: false,
        agent_id: null,
        issuer: null,
        capabilities: null,
        constraints: null,
        delegation_verified: null,
        delegation_chain: null,
        key_pinning: null,
        error_code: code,
        error_message: message,
        warnings: [],
    };
}

/**
 * Verify a credential offline using caller-provided documents.
 * Implements the 12-step verification flow from the spec.
 * @param {string} credentialJwt
 * @param {object} discovery - Discovery document
 * @param {object|null} revocation - Revocation document
 * @param {import('./pinning.js').KeyPinStore} pinStore
 * @param {string|null} audience
 * @param {object} config - Verifier config
 * @returns {object} VerificationResult
 */
export function verifyCredentialOffline(credentialJwt, discovery, revocation, pinStore, audience, config) {
    config = config || defaultVerifierConfig();

    // Step 1: Parse JWT (validates alg == ES256 and typ == agentpin-credential+jwt)
    let header, payload;
    try {
        ({ header, payload } = decodeJwtUnverified(credentialJwt));
    } catch (e) {
        return failureResult(ErrorCode.ALGORITHM_REJECTED, `JWT parse failed: ${e.message}`);
    }

    // Step 2: Check temporal validity
    const now = Math.floor(Date.now() / 1000);
    const skew = config.clockSkewSecs;

    if (payload.iat > now + skew) {
        return failureResult(ErrorCode.CREDENTIAL_EXPIRED, 'Credential issued in the future');
    }
    if (payload.exp <= now - skew) {
        return failureResult(ErrorCode.CREDENTIAL_EXPIRED, 'Credential has expired');
    }
    if (payload.nbf !== undefined && payload.nbf !== null && payload.nbf > now + skew) {
        return failureResult(ErrorCode.CREDENTIAL_EXPIRED, 'Credential not yet valid (nbf)');
    }

    // Check max TTL
    const lifetime = payload.exp - payload.iat;
    if (lifetime > config.maxTtlSecs) {
        return failureResult(
            ErrorCode.CREDENTIAL_EXPIRED,
            `Credential lifetime ${lifetime} exceeds max TTL ${config.maxTtlSecs}`
        );
    }

    // Step 3: Validate discovery document (entity matches iss)
    try {
        validateDiscoveryDocument(discovery, payload.iss);
    } catch (e) {
        return failureResult(ErrorCode.DISCOVERY_INVALID, `Discovery validation failed: ${e.message}`);
    }

    // Step 4: Resolve public key by kid
    const jwk = findKeyByKid(discovery, header.kid);
    if (!jwk) {
        return failureResult(ErrorCode.KEY_NOT_FOUND, `Key '${header.kid}' not found in discovery document`);
    }

    // Check key expiration
    if (jwk.exp) {
        const expDate = new Date(jwk.exp);
        if (expDate.getTime() / 1000 < now - skew) {
            return failureResult(ErrorCode.KEY_EXPIRED, `Key '${header.kid}' has expired`);
        }
    }

    // Convert JWK to PEM for verification
    let publicKeyPem;
    try {
        publicKeyPem = jwkToPem(jwk);
    } catch (e) {
        return failureResult(ErrorCode.KEY_NOT_FOUND, `Invalid key format for '${header.kid}': ${e.message}`);
    }

    // Step 5: Verify JWT signature
    try {
        verifyJwt(credentialJwt, publicKeyPem);
    } catch {
        return failureResult(ErrorCode.SIGNATURE_INVALID, `JWT signature verification failed for kid '${header.kid}'`);
    }

    // Step 6: Check revocation
    if (revocation) {
        try {
            checkRevocation(revocation, payload.jti, payload.sub, header.kid);
        } catch (e) {
            const code = e instanceof AgentPinError ? e.code : ErrorCode.CREDENTIAL_REVOKED;
            return failureResult(code, e.message);
        }
    }

    // Step 7: Validate agent status
    const agent = findAgentById(discovery, payload.sub);
    if (!agent) {
        return failureResult(ErrorCode.AGENT_NOT_FOUND, `Agent '${payload.sub}' not found in discovery document`);
    }
    if (agent.status !== 'active') {
        return failureResult(ErrorCode.AGENT_INACTIVE, `Agent '${payload.sub}' status is ${agent.status}`);
    }

    // Step 8: Validate capabilities
    try {
        validateCredentialAgainstDiscovery(payload.capabilities, agent.capabilities);
    } catch (e) {
        return failureResult(ErrorCode.CAPABILITY_EXCEEDED, e.message);
    }

    // Step 9: Validate constraints
    if (!constraintsSubsetOf(agent.constraints, payload.constraints)) {
        return failureResult(
            ErrorCode.CONSTRAINT_VIOLATION,
            'Credential constraints are less restrictive than discovery defaults'
        );
    }

    // Step 10: Validate delegation chain (offline mode: note only)
    let result = successResult(
        payload.sub,
        payload.iss,
        payload.capabilities,
        payload.constraints,
        { status: 'unknown', first_seen: null }
    );

    if (payload.delegation_chain) {
        const entries = payload.delegation_chain.map(att => ({
            domain: att.domain,
            role: att.role,
            verified: false,
        }));
        result.delegation_chain = entries;
        result.delegation_verified = false;
        result.warnings.push('Delegation chain present but not verified in offline mode');
    }

    // Step 11: TOFU key pinning
    try {
        const pinResult = checkPinning(pinStore, payload.iss, jwk);
        if (pinResult === PinningResult.FIRST_USE) {
            result.key_pinning = {
                status: 'first_use',
                first_seen: new Date().toISOString(),
            };
        } else if (pinResult === PinningResult.MATCHED) {
            const domainData = pinStore.getDomain(payload.iss);
            const firstSeen = domainData?.pinned_keys?.[0]?.first_seen || null;
            result.key_pinning = {
                status: 'pinned',
                first_seen: firstSeen,
            };
        }
    } catch {
        return failureResult(
            ErrorCode.KEY_PIN_MISMATCH,
            `Key for '${payload.iss}' has changed since last pinned`
        );
    }

    // Step 12: Check audience
    if (audience && payload.aud) {
        if (payload.aud !== '*' && payload.aud !== audience) {
            return failureResult(
                ErrorCode.AUDIENCE_MISMATCH,
                `Credential audience '${payload.aud}' does not match verifier '${audience}'`
            );
        }
    }

    return result;
}

/**
 * Online verification that fetches discovery/revocation documents.
 * @param {string} credentialJwt
 * @param {import('./pinning.js').KeyPinStore} pinStore
 * @param {string|null} audience
 * @param {object} config
 * @returns {Promise<object>}
 */
export async function verifyCredential(credentialJwt, pinStore, audience, config) {
    config = config || defaultVerifierConfig();

    // Parse JWT to extract issuer domain
    let payload;
    try {
        ({ payload } = decodeJwtUnverified(credentialJwt));
    } catch (e) {
        return failureResult(ErrorCode.ALGORITHM_REJECTED, `JWT parse failed: ${e.message}`);
    }

    // Fetch discovery document
    let discovery;
    try {
        discovery = await fetchDiscoveryDocument(payload.iss);
    } catch (e) {
        return failureResult(ErrorCode.DISCOVERY_FETCH_FAILED, `Failed to fetch discovery document: ${e.message}`);
    }

    // Fetch revocation document
    let revocation = null;
    if (discovery.revocation_endpoint) {
        try {
            revocation = await fetchRevocationDocument(discovery.revocation_endpoint);
        } catch {
            return failureResult(
                ErrorCode.DISCOVERY_FETCH_FAILED,
                'Revocation endpoint unreachable (fail-closed)'
            );
        }
    }

    return verifyCredentialOffline(credentialJwt, discovery, revocation, pinStore, audience, config);
}

File: python/agentpin/verification.py
"""Credential verification for AgentPin.

Implements the 12-step verification flow from the spec.
"""

import math
import time
from datetime import datetime, timezone
from typing import Optional

from .constraint import constraints_subset_of
from .credential import validate_credential_against_discovery
from .discovery import fetch_discovery_document, find_agent_by_id, find_key_by_kid, validate_discovery_document
from .jwk import jwk_to_pem
from .jwt import decode_jwt_unverified, verify_jwt
from .pinning import KeyPinStore, PinningResult, check_pinning
from .revocation import check_revocation, fetch_revocation_document
from .types import AgentPinError, ErrorCode, VerificationResult, VerifierConfig


def _success(
    agent_id: str,
    issuer: str,
    capabilities: list,
    constraints: Optional[dict],
    pin_status: dict,
) -> VerificationResult:
    return VerificationResult(
        valid=True,
        agent_id=agent_id,
        issuer=issuer,
        capabilities=capabilities,
        constraints=constraints,
        key_pinning=pin_status,
    )


def _failure(code: str, message: str) -> VerificationResult:
    return VerificationResult(valid=False, error_code=code, error_message=message)


def verify_credential_offline(
    credential_jwt: str,
    discovery: dict,
    revocation: Optional[dict],
    pin_store: KeyPinStore,
    audience: Optional[str],
    config: Optional[VerifierConfig] = None,
) -> VerificationResult:
    """Verify a credential offline using caller-provided documents.

    Implements the 12-step verification flow from the spec.
    """
    if config is None:
        config = VerifierConfig()

    # Step 1: Parse JWT
    try:
        header, payload, _sig = decode_jwt_unverified(credential_jwt)
    except Exception as e:
        return _failure(ErrorCode.ALGORITHM_REJECTED, f"JWT parse failed: {e}")

    # Step 2: Check temporal validity
    now = math.floor(time.time())
    skew = config.clock_skew_secs

    if payload.get("iat", 0) > now + skew:
        return _failure(ErrorCode.CREDENTIAL_EXPIRED, "Credential issued in the future")
    if payload.get("exp", 0) <= now - skew:
        return _failure(ErrorCode.CREDENTIAL_EXPIRED, "Credential has expired")
    nbf = payload.get("nbf")
    if nbf is not None and nbf > now + skew:
        return _failure(ErrorCode.CREDENTIAL_EXPIRED, "Credential not yet valid (nbf)")

    lifetime = payload.get("exp", 0) - payload.get("iat", 0)
    if lifetime > config.max_ttl_secs:
        return _failure(
            ErrorCode.CREDENTIAL_EXPIRED,
            f"Credential lifetime {lifetime} exceeds max TTL {config.max_ttl_secs}",
        )

    # Step 3: Validate discovery document
    try:
        validate_discovery_document(discovery, payload["iss"])
    except AgentPinError as e:
        return _failure(ErrorCode.DISCOVERY_INVALID, f"Discovery validation failed: {e}")

    # Step 4: Resolve public key by kid
    jwk = find_key_by_kid(discovery, header["kid"])
    if not jwk:
        return _failure(ErrorCode.KEY_NOT_FOUND, f"Key '{header['kid']}' not found in discovery document")

    # Check key expiration
    if jwk.get("exp"):
        try:
            exp_dt = datetime.fromisoformat(jwk["exp"])
            if exp_dt.tzinfo is None:
                exp_dt = exp_dt.replace(tzinfo=timezone.utc)
            if exp_dt.timestamp() < now - skew:
                return _failure(ErrorCode.KEY_EXPIRED, f"Key '{header['kid']}' has expired")
        except (ValueError, TypeError):
            pass

    # Convert JWK to PEM
    try:
        public_key_pem = jwk_to_pem(jwk)
    except Exception as e:
        return _failure(ErrorCode.KEY_NOT_FOUND, f"Invalid key format for '{header['kid']}': {e}")

    # Step 5: Verify JWT signature
    try:
        verify_jwt(credential_jwt, public_key_pem)
    except Exception:
        return _failure(ErrorCode.SIGNATURE_INVALID, f"JWT signature verification failed for kid '{header['kid']}'")

    # Step 6: Check revocation
    if revocation:
        try:
            check_revocation(revocation, payload["jti"], payload["sub"], header["kid"])
        except AgentPinError as e:
            return _failure(e.code, str(e))

    # Step 7: Validate agent status
    agent = find_agent_by_id(discovery, payload["sub"])
    if not agent:
        return _failure(ErrorCode.AGENT_NOT_FOUND, f"Agent '{payload['sub']}' not found in discovery document")
    if agent.get("status") != "active":
        return _failure(ErrorCode.AGENT_INACTIVE, f"Agent '{payload['sub']}' status is {agent.get('status')}")

    # Step 8: Validate capabilities
    try:
        validate_credential_against_discovery(payload.get("capabilities", []), agent.get("capabilities", []))
    except AgentPinError as e:
        return _failure(ErrorCode.CAPABILITY_EXCEEDED, str(e))

    # Step 9: Validate constraints
    if not constraints_subset_of(agent.get("constraints"), payload.get("constraints")):
        return _failure(
            ErrorCode.CONSTRAINT_VIOLATION,
            "Credential constraints are less restrictive than discovery defaults",
        )

    # Step 10: Delegation chain (offline: note only)
    result = _success(
        payload["sub"],
        payload["iss"],
        payload.get("capabilities", []),
        payload.get("constraints"),
        {"status": "unknown", "first_seen": None},
    )

    if payload.get("delegation_chain"):
        entries = [
            {"domain": att["domain"], "role": att["role"], "verified": False}
            for att in payload["delegation_chain"]
        ]
        result.delegation_chain = entries
        result.delegation_verified = False
        result.warnings.append("Delegation chain present but not verified in offline mode")

    # Step 11: TOFU key pinning
    try:
        pin_result = check_pinning(pin_store, payload["iss"], jwk)
        if pin_result == PinningResult.FIRST_USE:
            result.key_pinning = {"status": "first_use", "first_seen": datetime.now(timezone.utc).isoformat()}
        elif pin_result == PinningResult.MATCHED:
            domain_data = pin_store.get_domain(payload["iss"])
            first_seen = None
            if domain_data and domain_data.get("pinned_keys"):
                first_seen = domain_data["pinned_keys"][0].get("first_seen")
            result.key_pinning = {"status": "pinned", "first_seen": first_seen}
    except AgentPinError:
        return _failure(ErrorCode.KEY_PIN_MISMATCH, f"Key for '{payload['iss']}' has changed since last pinned")

    # Step 12: Check audience
    if audience and payload.get("aud"):
        cred_aud = payload["aud"]
        if cred_aud != "*" and cred_aud != audience:
            return _failure(
                ErrorCode.AUDIENCE_MISMATCH,
                f"Credential audience '{cred_aud}' does not match verifier '{audience}'",
            )

    return result


def verify_credential(
    credential_jwt: str,
    pin_store: KeyPinStore,
    audience: Optional[str],
    config: Optional[VerifierConfig] = None,
) -> VerificationResult:
    """Online verification that fetches discovery/revocation documents."""
    if config is None:
        config = VerifierConfig()

    # Parse JWT to extract issuer domain
    try:
        _header, payload, _sig = decode_jwt_unverified(credential_jwt)
    except Exception as e:
        return _failure(ErrorCode.ALGORITHM_REJECTED, f"JWT parse failed: {e}")

    # Fetch discovery document
    try:
        discovery = fetch_discovery_document(payload["iss"])
    except Exception as e:
        return _failure(ErrorCode.DISCOVERY_FETCH_FAILED, f"Failed to fetch discovery document: {e}")

    # Fetch revocation document
    revocation = None
    rev_endpoint = discovery.get("revocation_endpoint")
    if rev_endpoint:
        try:
            revocation = fetch_revocation_document(rev_endpoint)
        except Exception:
            return _failure(ErrorCode.DISCOVERY_FETCH_FAILED, "Revocation endpoint unreachable (fail-closed)")

    return verify_credential_offline(credential_jwt, discovery, revocation, pin_store, audience, config)
